### DISCUSSION TITLE
Configure Vaultwarden behind Nginx with Nginx Proxy Manager

### DISCUSSION THREAD

**Discussion by martymarty004** [StrausFuenf/hassio-addons#3] [Q&A] [answered]
Hello, I'm trying to set up Vaultwarden to use the fail2ban plugin, but I have issues with the configuration of the reverse proxy (I'm using the [NPM Add-on](https://github.com/hassio-addons/addon-nginx-proxy-manager) available in the Official Add-ons Repository).

The normal setup works without issues, but I don't understand how to configure the x-forwarded-for header to let Vaultwarden know the real IP of the client (now it only sees Nginx as the origin of the requests).

Example: 
`[2023-12-29 12:20:34.211][vaultwarden::api::identity][ERROR] Username or password is incorrect. Try again. IP: 172.30.32.1. Username: test@gmail.com`

This is the active config:
![1](https://github.com/StrausFuenf/hassio-addons/assets/61474670/4d377b2d-344d-4298-b66f-d124b476cdfc)
![2](https://github.com/StrausFuenf/hassio-addons/assets/61474670/37b2e649-1500-4aa0-956d-2e07e59a821b)
![3](https://github.com/StrausFuenf/hassio-addons/assets/61474670/9bf6f7e3-948e-4e55-a9f2-f027927208be)
![4](https://github.com/StrausFuenf/hassio-addons/assets/61474670/73d79089-3134-4cf4-bf7d-8b0487efee1b)

![image](https://github.com/StrausFuenf/hassio-addons/assets/61474670/24c63147-1db1-47a6-b788-1468adfca9d3)

---
By looking at the [Nginx Examples](https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples) in the official repository, I think that I'm missing the following piece of configuration:
```
location / {
      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;

      proxy_pass http://homeassistant:7277;
    }
```

I tried putting it in the location configuration but nothing, same result when I tried the Advanced tab.
Also, I think that it shouldn't be that necessary, because by looking at the actual configuration file that is being generated those headers are already there:

```txt
# cat /mnt/data/supervisor/addons/data/a0d7b954_nginxproxymanager/nginx/proxy_host/6.conf
# ------------------------------------------------------------
# bw.MYDOMAIN.TLD
# ------------------------------------------------------------

server {
  set $forward_scheme http;
  set $server         "homeassistant";
  set $port           7277;

  listen 80;
#listen [::]:80;

listen 443 ssl http2;
#listen [::]:443;

  server_name bw.MYDOMAIN.TLD;

  # Let's Encrypt SSL
  include conf.d/include/letsencrypt-acme-challenge.conf;
  include conf.d/include/ssl-ciphers.conf;
  ssl_certificate /etc/letsencrypt/live/npm-2/fullchain.pem;
  ssl_certificate_key /etc/letsencrypt/live/npm-2/privkey.pem;

  # Block Exploits
  include conf.d/include/block-exploits.conf;

  # HSTS (ngx_http_headers_module is required) (63072000 seconds = 2 years)
  add_header Strict-Transport-Security "max-age=63072000; preload" always;

    # Force SSL
    include conf.d/include/force-ssl.conf;

proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $http_connection;
proxy_http_version 1.1;

  access_log /proc/1/fd/1 proxy;
  error_log /proc/1/fd/1 warn;



  location / {
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-Scheme $scheme;
    proxy_set_header X-Forwarded-Proto  $scheme;
    proxy_set_header X-Forwarded-For    $remote_addr;
    proxy_set_header X-Real-IP          $remote_addr;
    proxy_pass       http://homeassistant:7277;

  # Block Exploits
  include conf.d/include/block-exploits.conf;

    # Force SSL
    include conf.d/include/force-ssl.conf;

  # HSTS (ngx_http_headers_module is required) (63072000 seconds = 2 years)
  add_header Strict-Transport-Security "max-age=63072000; preload" always;

    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $http_connection;
    proxy_http_version 1.1;
  }

  # Custom
  include /data/nginx/custom/server_proxy[.]conf;
}
```

Does anyone have any suggestions?

**Comment by strausfuenf** [comment]
Looks like there is a problem with your nginx setup.
I tried to recreate your problem with a fresh npm and vaultwarden.
I always get the right IP.
I even tried a new NPM on another machine and it worked..

**Reply by martymarty004** [reply]
@StrausFuenf are you able to show me your NPM configuration? Maybe I made some kind of mistake that I can't see...

Just in case, I will try a fresh install of the NPM add-on to see how it goes.

**Reply by martymarty004** [reply, marked as answer]
@StrausFuenf, I did some more tests and I think I found the issue.
When accessing Vaultwarden from the mobile network or the WireGuard tunnel, the origin IP is detected as expected. 
Even on some other networks, I don't face any issues.

Apparently, everything stops working if I connect **through IPv6** (my network is dual stack).
If the connection is made from an IPv6 address, then NPM shows `172.30.32.1` as the origin, which I think is the hassio bridge.
If I disable IPv6 on my client and force it to use IPv4, then the IP origin is correct.

I don't know if this is a bug with NPM, or an incompatibility with the docker network and IPv6.
I couldn't find any clue on the internet about it, so I'm leaving a trace here....

EDIT: I finally found a mention of the issue on an [issue](https://github.com/docker/roadmap/issues/157#issuecomment-726316121) in the docker repository. Apparently, this is expected behavior if IPv6 is not enabled in the docker network.

**Reply by strausfuenf** [reply]
Sorry was a bit busy. Normally im not using nginx proxy manager i just configure nginx sites to work as a proxy yeah looks like a docker/hassos problem. Maybe it would help to try and set docker ipv6 ,disable the dualstack in your network or fail2ban. Ill look if there is a workaround for that.

**Reply by strausfuenf** [reply]
Hey,
I testet some things again.
There would be an option to try to setup another addon which enables ipv6nat for docker.. But thats nothing i would try or implement  since it changes some docker network settings so thats a no for me.

You could add your Domain to cloudflare and run it behind the cloudflare proxy so it is Cloudflareproxy->NPM->Vaultwarden this allows you to remove X-Forwarded-For and use CF-Connecting-IP.

Hope that helps

/StrausFuenf

**Reply by martymarty004** [reply]
Yeah, using the Cloudflare proxy seems like the best option.
I was already thinking about buying a custom domain to use for my projects (right now I'm using DuckDNS).

The idea was to close all ports on my router and just configure a Cloudflared Tunnel for remote access. 
That way I will be able to hide my IP (which is static) and also take advantage of the Cloudflare WAF and other functionalities.
